.github/workflows/build-images-from-branch.yml

```diff
@@ -21,19 +21,81 @@ on:
       - 'develop'
 
 jobs:
-  push:
+  base:
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v2
+        if: github.event_name == 'workflow_dispatch'
+        with:
+          ref: '${{ github.event.inputs.branch_name }}'
+
+      - uses: actions/checkout@v2
+        if: github.event_name == 'push'
+
+      - name: Set version
+        id: set-version
+        run: |
+          export VERSION=`git log -1 --pretty=format:%h`
+          echo "version=$VERSION" >> $GITHUB_OUTPUT
+      - name: Get list of services
+        id: get-services
+        run: |
+          services=$(grep "^  [^ ]" docker-compose.yml | grep -v '#' | awk -F: '{print $1}' | sed -e 's/^ *//')
+          services_json=$(echo $services | tr '\n' ',' | sed 's/,$//' | jq -R 'split(" ")' | tr -d '\n')
+          # Set the list of service names as an output variable
+          echo "services=$services_json" >> $GITHUB_OUTPUT
+          echo "services=$services_json"
+      - name: Build base image
+        id: base-image
+        run: |
+          yarn build:image
+
+      - name: Tag image with version
+        run: |
+          docker image tag opencrvs-build:latest opencrvs-build:${{ steps.set-version.outputs.version }}
+
+      - name: Upload image
+        uses: ishworkh/docker-image-artifact-upload@v1
+        with:
+          image: 'opencrvs-build:${{ steps.set-version.outputs.version }}'
+
+    outputs:
+      services: ${{ steps.get-services.outputs.services }}
+      version: ${{ steps.set-version.outputs.version }}
+
+  build:
+    needs: base
+    strategy:
+      matrix:
+        service: ${{ fromJSON(needs.base.outputs.services) }}
     runs-on: ubuntu-latest
     steps:
       - uses: actions/checkout@v2
         if: github.event_name == 'workflow_dispatch'
         with:
           ref: '${{ github.event.inputs.branch_name }}'
+
       - uses: actions/checkout@v2
         if: github.event_name == 'push'
+
       - name: Login to DockerHub
         uses: docker/login-action@v1
         with:
           username: ${{ secrets.DOCKER_USERNAME }}
           password: ${{ secrets.DOCKER_PASSWORD }}
-      - name: Push image
-        run: yarn compose:push:version
+
+      - name: Download image
+        uses: ishworkh/docker-image-artifact-download@v1
+        with:
+          image: 'opencrvs-build:${{ needs.base.outputs.version }}'
+
+      - name: Rename image to latest
+        run: |
+          docker image tag opencrvs-build:${{ needs.base.outputs.version }} opencrvs-build:latest
+
+      - name: Build ${{ matrix.service }}
```

Comment:
@rikukissa what is matrix.service?

Reply:
@euanmillar it's the name of the service being built. This build strategy now builds each service at the same time on different build machines. "Matrix" in this context means the set of parallel build processes. The variables are defined here: https://github.com/opencrvs/opencrvs-core/pull/4752/files/7eac54d2a5a95f30067ee3b88279751dd13f5bba#diff-296935306c9732c880484facf16f9180b7291eed77275111c1791092d8ea70e4R62-R70